### ISSUE TITLE
B::Deparse and representation of infinity

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3949]
Migrated from [rt.perl.org#6944](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6944) (status was 'resolved')

Searchable as RT6944$

**Comment by p5pRT** [comment]
### From pnewton@gmx.de



B&#8203;::Deparse gave wrong results for me in constants which are calculated at
compile time and evaluate to infinity\.

Two examples with ActivePerl build 623 \(roughly Perl 5\.6\.0\)&#8203;:

U&#8203;:\\pne>perl \-MO=Deparse \-e "$a = 2\*\*4\.2\*\*12"
$a = 1\.\#INF;
\-e syntax OK

U&#8203;:\\pne>perl \-MO=Deparse\,\-p \-e "$a = 2\*\*4\.2\*\*12"
\($a = 1\.\#INF\);
\-e syntax OK

However\, what B&#8203;::Deparse spits out is not the same as what went in \-\- and
the second output doesn't even compile\, since the closing parenthesis is
inside a comment&#8203;:

U&#8203;:\\pne>perl \-MO=Deparse\,\-p \-e "$a = 1\.\#INF"
\($a = 1\);
\-e syntax OK

U&#8203;:\\pne>perl \-MO=Deparse\,\-p \-e "$a = \(1\.\#INF\);"
syntax error at \-e line 1\, at EOF
\-e had compilation errors\.
;


Since I know that B&#8203;::Deparse has been changed a lot recently\, I also tried
this out with perl@&#8203;9945 on Linux\. The results are similar\.

\[pne@&#8203;penderel perl\]$ \./perl \-Ilib \-MO=Deparse \-e '$a = 2\*\*4\.2\*\*12'
$a = inf;
\-e syntax OK
\[pne@&#8203;penderel perl\]$ \./perl \-Ilib \-MO=Deparse\,\-p \-e '$a = 2\*\*4\.2\*\*12'
\($a = inf\);
\-e syntax OK

In this case\, both outputs are syntactically\, but not semantically\,
correct \(they use the bareword 'inf'\)\.

Also interesting was the following&#8203;:

\[pne@&#8203;penderel perl\]$ \./perl \-Ilib \-de1
Default die handler restored\.

Loading DB routines from perl5db\.pl version 1\.07
Editor support available\.

Enter h or \`h h' for help\, or \`man perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> x 2\*\*4\.2\*\*12
0  4294967295
&nbsp;  DB\<2> p 2\*\*4\.2\*\*12
inf
&nbsp;  DB\<3> print 2\*\*4\.2\*\*12
inf


Where 'x' in the debugger yielded MAXINT \(in the main Perl on my path\,
it's 2\*\*64\-1 rather than 2\*\*32\-1\, but the idea is the same\)\, but both
'p' and 'print' gave "inf"\.

In ActivePerl\, the debugger session looks like this&#8203;:

main&#8203;::\(\-e&#8203;:1\)&#8203;:   1
&nbsp;  DB\<1> x 2\*\*4\.2\*\*12
0  '1\.\#INF'
&nbsp;  DB\<2> p 2\*\*4\.2\*\*12
1\.\#INF
&nbsp;  DB\<3> print 2\*\*4\.2\*\*12
1\.\#INF


Which is nearly the same\.

Well\, I suppose B&#8203;::Deparse can't do a lot if the calculation has already
been constant folded by Perl\, but just saying 1\.\#INF or inf doesn't work :\)

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.6.0:

Configured by pne at Fri Dec 15 16:26:56 2000.

Summary of my perl5 (revision 5 version 6 subversion 0) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=undef d_sfio=undef uselargefiles=undef 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cl', optimize='-O1 -MD -DNDEBUG', gccversion=
    cppflags='-DWIN32'
    ccflags ='-O1 -MD -DNDEBUG -DWIN32 -D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -DPERL_IMPLICIT_CONTEXT -
DPERL_IMPLICIT_SYS -DPERL_MSVCRT_READFIX'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release  -libpath:"C:\Perl\lib\CORE"  -machine:x86'
    libpth="C:\Perl\lib\CORE"
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  
netapi32.lib uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl56.lib
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release  -libpath:"C:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY


@INC for perl v5.6.0:
    C:/Perl/lib
    C:/Perl/site/lib
    .


Environment for perl v5.6.0:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Perl\bin\;C:\bin;C:\orant\bin;c:\dmi\win32\bin;C:\WINNT\system32;C:\WINNT;C:\Programme\Network 
Associates\PGPNT;C:\PROGRA~1\Tcl\bin;C:\Programme\ActivePython\
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
@cwest - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From pnewton@gmx.de

On 6 Jul 2003 at 22&#8203;:58\, Casey West via RT wrote&#8203;:

> According to our records\, your request regarding 
>   "B&#8203;::Deparse and representation of infinity" 
> has been resolved\. 

How so?

I realise the request was for 5\.6\.0\, which is probably not supported\.

However\, rsyncing the latest bleadperl and testing with that shows that
the behaviour is still there\, at least on some platforms&#8203;:

bash\-2\.05b$ \./perl \-Ilib \-MO=Deparse \-e '$a = 2\*\*4\.2\*\*12'
$a = inf;
\-e syntax OK

This is the perl I used&#8203;:

bash\-2\.05b$ \./perl \-Ilib \-V
Summary of my perl5 \(revision 5\.0 version 9 subversion 0 patch 20043\) 
configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.2\.19\, archname=i586\-linux
&nbsp;    uname='linux penderel 2\.2\.19 \#1 sat jun 9 13&#8203;:04&#8203;:06 est 2001 i586 
unknown unknown gnulinux '
&nbsp;    config\_args='\-Duselargefiles \-Dusedevel'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include \-
D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O3'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='2\.95\.4 20011002 \(Debian prerelease\)'\, 
gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=/lib/libc\-2\.3\.1\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.1'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-
rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;:
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Locally applied patches&#8203;:
&nbsp;        DEVEL18374
&nbsp;  Built under linux
&nbsp;  Compiled at Jul  7 2003 10&#8203;:11&#8203;:39
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    lib
&nbsp;    /usr/local/lib/perl5/5\.9\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/5\.9\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.9\.0/i586\-linux
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.9\.0
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.

**Comment by p5pRT** [comment]
### From smcc@mit.edu

I believe I was able to fix this in version 0\.64 of B&#8203;::Deparse\, so it
should work in 5\.8\.1 and 5\.10\. Deparse now prints infinite values as
an expression that will evaluate to an infinite value on just about any
platform&#8203;:

% perl \-MO=Deparse \-e '$a = 2\*\*4\.2\*\*12'
$a = 9\*\*9\*\*9;

**Comment by p5pRT** [comment]
### From @steve-m-hay

This is indeed now fixed as per the previous correspondence&#8203;:

C&#8203;:\\Temp>E&#8203;:\\Perls\\perl581\\bin\\perl \-MO=Deparse \-e "$a = 2\*\*4\.2\*\*12"
$a = 9\*\*9\*\*9;
\-e syntax OK

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
